File: Dsa-day174.cpp
class Solution {
private:
    long long maxProd = 0;
    long long totalSum = 0;
    const int MOD = 1000000007;
    
    long long calculateSum(TreeNode* node) {
        if (node == nullptr) return 0;
        return node->val + calculateSum(node->left) + calculateSum(node->right);
    }
    
    long long findMaxProduct(TreeNode* node) {
        if (node == nullptr) return 0;
        
        long long leftSum = findMaxProduct(node->left);
        long long rightSum = findMaxProduct(node->right);
        long long currentSum = node->val + leftSum + rightSum;
        
        long long product = currentSum * (totalSum - currentSum);
        maxProd = max(maxProd, product);
        
        return currentSum;
    }
    
public:
    int maxProduct(TreeNode* root) {
        totalSum = calculateSum(root);
        findMaxProduct(root);
        return maxProd % MOD;
    }
};

File: Dsa-day174.py
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        self.maxProduct = 0
        self.totalSum = 0
        MOD = 10**9 + 7
        
        def calculateSum(node):
            if not node:
                return 0
            return node.val + calculateSum(node.left) + calculateSum(node.right)
        
        def findMaxProduct(node):
            if not node:
                return 0
            
            leftSum = findMaxProduct(node.left)
            rightSum = findMaxProduct(node.right)
            currentSum = node.val + leftSum + rightSum
            
            product = currentSum * (self.totalSum - currentSum)
            self.maxProduct = max(self.maxProduct, product)
            
            return currentSum
        
        self.totalSum = calculateSum(root)
        findMaxProduct(root)
        return self.maxProduct % MOD